imgpkg/002-recursive-bundles/README.md

```diff
@@ -0,0 +1,328 @@
+copy | imported 4 images
+Succeeded
+```
+
+### From Repository to Repository
+
+#### Required changes
+
+##### Retrieve all images for all bundles
+
+The most significant change in this operation is that when we copy the images `imgpkg` will have to traverse all the bundles to collect all the images to copy.
+
+##### Bundles/Images deduped
+
+When copying the images/bundles ensure that `imgpkg` does not try to copy the same image multiple times
+
+#### Example
+
+Given that we create a bundle using the command 
+
+`imgpkg push -b ghcr.io/k14s/design-docs/simple-app-install-package -f imgpkg/002-recursive-bundles/examples/recursive-bundle`
+
+```=
+imgpkg copy -b ghcr.io/k14s/design-docs/simple-app-install-package --to-repo some-other.registry.io/recursive-bundle
+copy | exporting 4 images...
+copy | will export ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+copy | will export ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f
+copy | will export ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4
+copy | will export ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035
+copy | exported 4 images
+copy | importing 4 images...
+copy | importing ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0 -> some-other.registry.io/recursive-bundle@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0...
+copy | importing ghcr.io/k14s/design-docs/utilities@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f -> some-other.registry.io/recursive-bundle@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f...
+copy | importing ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4 -> some-other.registry.io/recursive-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4...
+copy | importing ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035 -> some-other.registry.io/recursive-bundle@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035...
+copy | imported 4 images
+Succeeded
+```
+
+## Pull a bundle
+
+### Required changes
+
+#### Change folder structure
+
+When using the `pull` command without the `-r` flag the behavior is not changed. Nothing will be dowloaded from the nested bundles.
+
+When downloading a bundle that contains other bundles to disk using the `pull -r` command will work as currently, except for that it will download the nested bundles into a hidden folder called `.bundles`.
```

Comment:
As a user, I think i'd be surprised that the recursive behavior is not the default.

And that I'd need to know if a bundle references other bundles, then add the `-r` flag.

Reply:
That in fact was the original proposal, but we updated it because we thought that it would give us a little more flexibility for possible future features.
If we think of a bundle to be a group of images and a configuration for an application the default behavior should be pulled recursively. But if we think about a bundle as a repository of applications you might not want to recursively pull all the applications to read this repository information.

To be fair we do not know yet what will be the more common use case for the recursive bundles. Given the information we have today we believe that making the default behavior not pull recursively is going to allow us to cater to possible future features more easily.

Reply:
I agree with Josh fwiw, I would consider non-recursive to the be exception vs. the norm.

Reply:
Given the current use case we have for Recursive Bundles (bundle as a repository of applications) we believe that making the default behavior not pull recursively would be more helpful for the users of `imgpkg`.
In the future, if we see a change in the usage of `imgpkg` we might change this behavior default behavior.